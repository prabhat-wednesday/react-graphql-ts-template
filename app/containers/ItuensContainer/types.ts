export interface Song {
  trackId: number;
  artistName: string;
  artworkUrl100: string;
  collectionName: string;
}

export interface SongData {
  Songs: Song[];
}
