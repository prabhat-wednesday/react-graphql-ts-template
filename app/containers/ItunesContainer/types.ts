import { AnyAction } from '@reduxjs/toolkit';

export interface Song {
  trackId: number;
  artistName: string;
  artworkUrl100: string;
  collectionName: string;
  previewUrl: string;
}

export interface SongData {
  results: Song[];
}

export type SongActionCreator = (payload: string) => AnyAction;

export interface ItuneContainerProps {
  dispatchArtistName: SongActionCreator;
  songData: {
    results: Song[];
  };
  loading: boolean;
  songListError?: string;
}
