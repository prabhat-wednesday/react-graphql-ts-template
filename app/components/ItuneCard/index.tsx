import React, { useRef, useEffect } from 'react';
import { Card } from 'antd';
import styled from 'styled-components';
import { media } from '@app/themes';
import If from '../If';

const { Meta } = Card;

export interface ItuneCardProps {
  trackId: number;
  artistName: string;
  artworkUrl100: string;
  collectionName: string;
  previewUrl: string;
  currentTrackId: number;
  handleOnPlay: (number: any) => any;
}

const CustomCard = styled(Card)`
  && {
    padding: 5px;
    margin: 1rem;
    width: 17rem;

    ${media.lessThan('tablet')`
      width: 12rem;
      margin: 0.5rem;
    `}
  }
`;

const StyledImg = styled.img`
  && {
    max-height: 15rem;
    object-fit: cover;
  }
`;

const StyledAudio = styled.audio`
  && {
    width: 100%;
    padding: 5px;
  }
`;

const StyledMeta = styled(Meta)`
  && {
    padding: 5px;
  }
`;

const ItuneCard = ({
  trackId,
  artistName,
  artworkUrl100,
  collectionName,
  previewUrl,
  currentTrackId,
  handleOnPlay
}: ItuneCardProps) => {
  const audioRef = useRef(null);
  useEffect(() => {
    if (audioRef.current) {
      if (currentTrackId !== trackId) {
        audioRef.current?.pause();
      }
    }
  }, [currentTrackId]);
  return (
    <If condition={trackId}>
      <CustomCard
        hoverable
        key={trackId}
        bodyStyle={{ padding: '10px' }}
        cover={<StyledImg src={artworkUrl100} loading="lazy" data-testid="cover-img" />}
      >
        <StyledMeta title={artistName} description={collectionName} />
        <StyledAudio
          data-testid="audio-element"
          onPlay={() => handleOnPlay(trackId)}
          ref={audioRef}
          controls
          src={previewUrl}
        >
          <a href={previewUrl}>Audio Preview</a>
        </StyledAudio>
      </CustomCard>
    </If>
  );
};

export default React.memo(ItuneCard);
