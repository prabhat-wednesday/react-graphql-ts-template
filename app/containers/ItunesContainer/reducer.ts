import { prepare } from '@app/utils';
import { createSlice } from '@reduxjs/toolkit';
import { song } from './types';

export interface ituneResponseState {
  loading: boolean;
  songData: {
    resultCount: number;
    results: song[];
  };
  songListError: string;
}

export const initialState: ituneResponseState = {
  loading: false,
  songData: { resultCount: 0, results: [] },
  songListError: ''
};

const ituneSlice = createSlice({
  name: 'Itune',
  initialState,
  reducers: {
    requestGetSongList: {
      reducer: (state) => {
        state.loading = true;
      },
      prepare
    },
    successGetSongList: (state, action) => {
      state.songData = action.payload;
      state.songListError = '';
      state.loading = false;
    },
    failureGetSongList: (state, action) => {
      state.songData.resultCount = 0;
      state.songData.results = [];
      state.loading = false;
      state.songListError = action.payload;
    }
  }
});

export const { requestGetSongList, successGetSongList, failureGetSongList } = ituneSlice.actions;

export default ituneSlice.reducer;
