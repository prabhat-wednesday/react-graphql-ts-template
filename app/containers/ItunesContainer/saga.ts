import { takeLatest, call, put } from 'redux-saga/effects';
import { requestGetSongList, successGetSongList, failureGetSongList } from '../SongProviderContainer/reducer';
import { getItune } from '@services/apiUtils';

export function* fetchDataFromItune(action: any): Generator<any, any> {
  const { artistName, pageNumber, pageSize } = action.payload ?? { artistName: '', pageNumber: 0, pageSize: 0 };
  const res: any = yield call(getItune, { artistName, pageNumber, pageSize });
  const { data, ok, error } = res;
  if (ok) {
    yield put(successGetSongList(data));
  } else {
    yield put(failureGetSongList(error));
  }
}

function* ituneCallSaga() {
  yield takeLatest(requestGetSongList.toString(), fetchDataFromItune);
}

export default ituneCallSaga;
