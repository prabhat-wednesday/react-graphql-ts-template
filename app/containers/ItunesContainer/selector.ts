import { createSelector } from 'reselect';
import get from 'lodash-es/get';
import { initialState } from './reducer';

const selectItuneState = (state: any) => state.ituneReducer || initialState;

export const selectLoading = () => createSelector(selectItuneState, (substate) => get(substate, 'loading'));
export const selectSongData = () => createSelector(selectItuneState, (substate) => get(substate, 'songData'));
export const selectSongListError = () => createSelector(selectItuneState, (substate) => get(substate, 'error'));
