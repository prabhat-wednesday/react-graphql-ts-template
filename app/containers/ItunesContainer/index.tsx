import React, { ChangeEvent, useEffect, useState } from 'react';
import debounce from 'lodash-es/debounce';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import { injectSaga } from 'redux-injectors';
import { AnyAction, compose } from '@reduxjs/toolkit';
import { selectLoading, selectSongData } from './selector';
import { requestGetSongList } from '../SongProviderContainer/reducer';
import ituneCallSaga from './saga';
import { ItuneContainerProps, RequestSongListActionPayload } from './types';
import ItuneSongList from '@app/components/ItuneSongList';
import styled from 'styled-components';
import { Input, Pagination, PaginationProps } from 'antd';
import { media } from '@app/themes';
import { If, T } from '@app/components';
import { useHistory } from 'react-router-dom';
import { setQueryParam } from '@app/utils';
import { isEmpty } from 'lodash-es';
import db from '../../db.json';
import CarouselContainer from '../CarouselContainer';
import { logoutUser } from '../LoginContainer/reducer';

const InputContainer = styled.div`
  && {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 0.5rem auto;
  }
`;

const CustomInput = styled(Input)`
  && {
    width: 30%;
    display: block;

    ${media.lessThan('tablet')`
    width: 80%;`}
  }
`;

const CustomPagination = styled(Pagination)`
  display: flex;
  align-items: center;
  justify-content: center;
`;

const ItunesContainer = ({ dispatchSongList, songData, loading, dispatchLogoutUser }: ItuneContainerProps) => {
  console.log(db);
  const { results } = songData;
  const [paginationParams, setPaginationParams] = useState({ pageNumber: 1, pageSize: 10 });
  const history = useHistory();
  const artistName = new URLSearchParams(history.location.search).get('artist_name');
  const setArtistName = (artistName: string) => setQueryParam({ param: 'artist_name', value: artistName });
  const { pageNumber, pageSize } = paginationParams;

  const handlePaginationOnChange: PaginationProps['onChange'] = (pageNumber: number, pageSize) => {
    if (pageNumber !== undefined) {
      setPaginationParams((prev) => ({ ...prev, pageNumber }));
    }
    if (pageSize !== undefined) {
      setPaginationParams((prev) => ({ ...prev, pageSize }));
    }
  };

  useEffect(() => {
    if (artistName !== undefined) {
      dispatchSongList({ artistName, pageNumber, pageSize });
    }
  }, [artistName, pageNumber, pageSize]);

  const handleOnChange = debounce((e: ChangeEvent<HTMLInputElement>) => {
    const artistName = e.target.value;
    if (artistName.trim()) {
      setArtistName(artistName);
      dispatchSongList({ artistName });
    }
  }, 500);
  return (
    <div>
      <InputContainer>
        <button onClick={dispatchLogoutUser}>Logout</button>
        <T data-testid="search-label" id="song_search_default" />
        <CustomInput
          data-testid="search-bar"
          aria-label="input-element"
          onChange={(e) => handleOnChange(e)}
          type="text"
        />
      </InputContainer>
      <CarouselContainer db={db} />
      <ItuneSongList loading={loading} songData={songData} />
      <If condition={!isEmpty(results)}>
        <CustomPagination onChange={handlePaginationOnChange} defaultCurrent={1} total={50} />
      </If>
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  loading: selectLoading(),
  songData: selectSongData()
});

export function mapDispatchToProps(dispatch: (arg0: AnyAction) => any) {
  return {
    dispatchSongList: (payload: RequestSongListActionPayload) => dispatch(requestGetSongList(payload)),
    dispatchLogoutUser: () => dispatch(logoutUser())
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect, injectSaga({ key: 'ituneComponent', saga: ituneCallSaga }))(ItunesContainer);

export const ItunesContainerTest = ItunesContainer;
