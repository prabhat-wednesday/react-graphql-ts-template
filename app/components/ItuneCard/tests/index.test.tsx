import React from 'react';
import ItuneCard from '..';
import { Song } from '@app/containers/ItunesContainer/types';
import { render, screen } from '@testing-library/react';

describe('<ItuneCard />', () => {
  let song: Song;

  beforeAll(() => {
    song = {
      trackId: 1,
      artistName: 'Yung Xiety',
      artworkUrl100:
        'https://is1-ssl.mzstatic.com/image/thumb/Music112/v4/3b/cc/31/3bcc315a-e58f-c678-8d05-b7d62346070f/1d724fe3-daaa-4825-91f0-a47e70a438bd.jpg/100x100bb.jpg',
      collectionName: 'Arijit Singh (Mashup) - Single'
    };
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = render(<ItuneCard {...song} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should render and match the snapshot when song object is empty', () => {
    const song = {};
    const { baseElement } = render(<ItuneCard {...song} />);
    expect(baseElement).toMatchSnapshot();
    expect(screen.queryByRole('img')).toBe(null);
    expect(screen.queryByRole('Yung Xiety')).toBe(null);
  });
});
