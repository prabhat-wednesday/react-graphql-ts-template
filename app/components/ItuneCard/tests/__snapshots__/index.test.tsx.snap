// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItuneCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered ant-card-hoverable ItuneCard__CustomCard-sc-1g3ywlw-0 YOMkc"
    >
      <div
        class="ant-card-cover"
      >
        <img
          class="ItuneCard__StyledImg-sc-1g3ywlw-1 dHbSaT"
          data-testid="cover-img"
          loading="lazy"
          src="https://is1-ssl.mzstatic.com/image/thumb/Music112/v4/3b/cc/31/3bcc315a-e58f-c678-8d05-b7d62346070f/1d724fe3-daaa-4825-91f0-a47e70a438bd.jpg/100x100bb.jpg"
        />
      </div>
      <div
        class="ant-card-body"
        style="padding: 10px;"
      >
        <div
          class="ant-card-meta ItuneCard__StyledMeta-sc-1g3ywlw-3 gbJaIc"
        >
          <div
            class="ant-card-meta-detail"
          >
            <div
              class="ant-card-meta-title"
            >
              Yung Xiety
            </div>
            <div
              class="ant-card-meta-description"
            >
              Arijit Singh (Mashup) - Single
            </div>
          </div>
        </div>
        <audio
          controls=""
          data-testid="audio-element"
          src="https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview122/v4/90/d6/0d/90d60d3b-2b95-8c84-f9aa-f69a2793d22c/mzaf_17815855687762407827.plus.aac.p.m4a"
        >
          <a
            href="https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview122/v4/90/d6/0d/90d60d3b-2b95-8c84-f9aa-f69a2793d22c/mzaf_17815855687762407827.plus.aac.p.m4a"
          >
            Audio Preview
          </a>
        </audio>
      </div>
    </div>
  </div>
</body>
`;

exports[`<ItuneCard /> should render and match the snapshot when song object is empty 1`] = `
<body>
  <div />
</body>
`;
